shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll

```diff
@@ -2985,29 +3019,20 @@ module MakeImpl<LocationSig Location, InputSig<Location> Lang> {
             or
             // flow into a callable
             exists(
-              ArgNodeEx arg, boolean allowsFlowThrough, Cc outercc, ParamNodeOption outerSummaryCtx,
-              TypOption outerArgT, ApOption outerArgAp
+              ArgNodeEx arg, boolean allowsFlowThrough, Cc outercc, SummaryCtx outerSummaryCtx
             |
-              pn1 = TPathNodeMid(arg, state, outercc, outerSummaryCtx, outerArgT, outerArgAp, t, ap) and
-              fwdFlowInStep(arg, node, state, outercc, cc, outerSummaryCtx, outerArgT, outerArgAp,
-                t, ap, allowsFlowThrough) and
+              pn1 = TPathNodeMid(arg, state, outercc, outerSummaryCtx, t, ap) and
+              fwdFlowInStep(arg, node, state, outercc, cc, outerSummaryCtx, t, ap, allowsFlowThrough) and
```

Comment:
I wonder if we should split `fwdFlowInStep` into two (`nomagic`'ed) predicates: `fwdFlowInStepMayFlowThrough` and `fwdFlowInStepMayNotFlowThrough`. That will make the non-linear recursion smaller, as we only need to join `fwdFlowInStepMayFlowThrough` against `TSummaryCtxSome`.

Reply:
As discussed offline: This isn't the place of the recursive call, but the question remains potentially relevant for the join in `fwdFlow0`. But lets consider that as a potential follow-up tweak.